### PR TITLE
1965 oembed

### PR DESCRIPTION
**Pull request by danicarrion**
controller expects a URL such as:

http://development.cartodb.local:3000/api/v1/oembed?format=json&maxwidth=50&url=http://development.cartodb.local:3000/viz/f9dbb21a-a89c-11e4-ae66-80fa5b110da5/

and will return something like:

    {
        type: "rich",
        version: "1.0",
        width: "100%",
        height: "100%",
        html: "<iframe width='100%' height='100%' frameborder='0'         src='http://development.cartodb.local:3000/viz/f9dbb21a-a89c-11e4-ae66-80fa5b110da5/embed_map' allowfullscreen webkitallowfullscreen mozallowfullscreen oallowfullscreen msallowfullscreen></iframe>"
    }

it complies with [oEmbed 1.0](http://oembed.com/)

both urls must refer to the same host and checks are run to ensure the URL param belong to an actual visualization and is valid

CR @juanignaciosl

### CODE REVIEW

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/1349/))

File: app/controllers/api/json/oembed_controller.rb [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+  include CartoDB
+
+  skip_before_filter :api_authorization_required
+
+  # Returns oembed data as required
+  def show
+    url = params[:url]
+    width = params[:maxwidth] || '100%'
+    height = params[:maxheight] || '100%'
+    format = request.query_parameters[:format]
```
Comment (juanignaciosl) [inline app/controllers/api/json/oembed_controller.rb]: Shouldn't this be `format = request.query_parameters[:format] || params[:format]` in order to retrieve ":json" for those cases without format param in query string?

File: app/controllers/api/json/oembed_controller.rb [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+    uri = URI.parse(url)
+
+    if uri.host != request.host
+      raise ActionController::RoutingError.new('URL origin not allowed')
+    end
+
+    uuid = uri.path.split('/')[-1]
+    begin
+      CartoDB::Visualization::Member.new(id: uuid, type: 'derived').fetch
```
Comment (juanignaciosl) [inline app/controllers/api/json/oembed_controller.rb]: I'm afraid I've told you one thing wrong: Member is not a Sequel model (Sequel allows this kind of queries), and `type: 'derived'` has no effect. You could do this:
```
vis = CartoDB::Visualization::Member.new(id: uuid).fetch
raise KeyError.new('not derived') if vis.type != 'derived'
vis
```

File: app/controllers/api/json/oembed_controller.rb [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    rescue KeyError
+      raise ActionController::RoutingError.new('Visualization not found: ' + uuid)
+    end
+
+    if public_visualizations_show_path(id: uuid) != uri.path && public_visualizations_show_path(id: uuid) + '/' != uri.path
```
Comment (juanignaciosl) [inline app/controllers/api/json/oembed_controller.rb]: For regexp fans, and you run `public_visualizations_show_path` only once without another variable:
```
raise ActionControler... if uri.path =~ /^#{public_visualizations_show_path(id: uuid)}[\/]?$/
```

Not important at all anyway (and some people would even dislike it), so feel free to ignore this comment :_)

File: app/controllers/api/json/oembed_controller.rb (hunk 1 above)
Comment (danicarrion) [inline app/controllers/api/json/oembed_controller.rb]: This case should be covered by the if clause at the bottom, which also copes with the case where no format is specified at all, no?

File: app/controllers/api/json/oembed_controller.rb (hunk 2 above)
Comment (danicarrion) [inline app/controllers/api/json/oembed_controller.rb]: so, I just need to remove the `type: 'derived'` parameter, correct?

File: app/controllers/api/json/oembed_controller.rb (hunk 3 above)
Comment (danicarrion) [inline app/controllers/api/json/oembed_controller.rb]: much more elegant, this is :)

**Comment by danicarrion** [comment]
I've also added the title of the visualization in the response, which I'd forgotten

**Comment by javisantana** [comment]
what about a regexp on width and height to avoid people insert custom html?

File: app/controllers/api/json/oembed_controller.rb (hunk 1 above)
Comment (juanignaciosl) [inline app/controllers/api/json/oembed_controller.rb]: Ok, you're right

**Comment by juanignaciosl** [comment]
Nothing else to say :+1:

**Comment by danicarrion** [comment]
maxwidth and maxheight should be allowed if we want to comply with oembed 1.0

"Responses of this type must obey the maxwidth and maxheight request parameters."

**Comment by danicarrion** [comment]
ah, sorry, you mean sanitizing height and width? well, it's *their* html anyway, isn't it? i mean, they can change the html client-side anyway...

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/1365/))

**Comment by danicarrion** [comment]
width and height sanitized

**Comment by javisantana** [comment]
retest this please

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/1367/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/1370/))

**Comment by danicarrion** [comment]
uhmmm, i just hate double "/" in urls, like when you concatenate "http://test.com/" and "/page1". File.join() handles that. I tried URI.join() but it didn't work. Is there any other way to do this in ruby?

**Comment by juanignaciosl** [comment]
:+1:

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/1371/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/1375/))

**Pull request merged**